perfaware/sim86/shared/Sim8086.cpp

```diff
@@ -44,13 +47,54 @@ enum Flags
     Flag_count
 };
 
-static void SetFlags(u16 Result, bool* FlagArray)
-{
-    FlagArray[Flag_SF] = Result & 0x8000;
-    FlagArray[Flag_ZF] = (Result == 0);
-}
+static void SetFlags(const instruction& Instruction, const u16 LeftOperandValue, 
+    const u16 RightOperandValue, const u16 Result, bool* FlagArray)
+{   
+    u16 HighOrderBit = (Instruction.Operands[1].Register.Count == 1) ? 0x80 : 0x8000;
+    bool Parity = true;
+       
+    switch (Instruction.Op)
+    {
+        case Op_add:
+        {
+            FlagArray[Flag_CF] = ((LeftOperandValue & HighOrderBit) || (RightOperandValue & HighOrderBit))
+                && !(Result & HighOrderBit);
+            FlagArray[Flag_AF] = ((LeftOperandValue & 0x8) || (RightOperandValue & 0x8))
+                && !(Result & 0x8);
+            FlagArray[Flag_OF] = (!((LeftOperandValue ^ RightOperandValue) & HighOrderBit))
+                && ((Result ^ LeftOperandValue) & HighOrderBit);
+            FlagArray[Flag_SF] = Result & HighOrderBit;
+            FlagArray[Flag_ZF] = (Result == 0);
+            for (u16 i = 0; i < 8; i++)
+            {
+                if (Result & (1 << i))
+                {
+                    Parity = !Parity;
+                }
+            }
+            FlagArray[Flag_PF] = Parity;
+        } break;
+        case Op_sub: [[fallthrough]];
+        default:
```

Comment:
I think that this is not a very safe way of writing this code. Default captures all the cases, and it never checks that the `Instruction.Op` is valid. It could be set to 100000 and it would still enter the default case without noticing that there is a problem.

It's likely acceptable if you're working alone on your own project and you know nobody is never come and write an instruction with a wrong Op. But you never know. At the very least, if you don't want to rearchitecture this bit of code, the default case should have an assert that the Op is valid, this is a very easy thing to do that can yield a lot more confidence that there are no nasty surprises.

Reply:
You're right. For what it's worth, this function can only be reached via Op_add, Op_sub or Op_cmp because there are safeguards earlier.
But these might be removed, so I put an assert here. If this was more than a personal project I'd put proper error handling.
3dd47bb7df7996fefafadf6528ffbbd2c60da158